### PR TITLE
Use urlencode() function for variable `return` in URL query string

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue #19130.

### Summary of Changes
Use `urlencode()` function for result of `base64_encode()` in URL query string.

For develop joomla website on URL like `http://localhost/~user/joomla`  function `base64_encode()` generates unsafe string in URL query.

### Testing Instructions
Test issue #19130

Additional:
1. Login as admin, go to some page (ex user login page) on front end and save/save and close/cancel module settings. After save + save and close/save and close/cancel you should back to previous page.

2. Go to category list and find the button "New" (add new article to category).
Click on link and then cancel, or save article. After you should back to category list.
Then try to edit some article from the category list. Save or cancel and back to category list.

### Expected result
The variable `return` is correct and you will be redirected to the previous page.

### Actual result
This is only visible on website that contains `~` in address.

The variable `return` is invalid and joomla redirects you to home page.

### Documentation Changes Required
None

### CODE REVIEW

**Comment by ghost** [comment]
@julienV can you please test as you reported Issue #19130?

**Comment by julienV** [comment]
Yes, i will test asap, maybe just not today...

**Comment by csthomas** [comment]
Is there a chance to test this week?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19196">issues.joomla.org/tracker/joomla-cms/19196</a>.</sub>

**Comment by julienV** [comment]
tested, works as expected

**Comment by julienV** [comment]
tested successfully<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19196">issues.joomla.org/tracker/joomla-cms/19196</a>.</sub>

**Comment by julienV** [comment]
I have tested this item :white_check_mark: successfully on 1e64a414c8566e5e43064669074e2130911c8832<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19196">issues.joomla.org/tracker/joomla-cms/19196</a>.</sub>

**Comment by Quy** [comment]
With/without PR, I followed the test instructions #19130 and it redirected to the home page instead of article category list page.

The `New` button URL has the `+`:
`/~user/joomla/component/content/?task=article.add&amp;return=aHR0cDovL2xvY2FsaG9zdC9+dXNlci9qb29tbGEvYXJ0aWNsZS1jYXRlZ29yeS1saXN0&amp;a_id=0&amp;catid=19`

**Comment by julienV** [comment]
well, worked fine with me the method with module editing: without patch, went back to index, with patch, went back to initial page.
are you sure you refreshed properly after applying the patch (i'm asking before it happened to me... patch, go back to index, then to another menu item, edit module, save and close)

**Comment by Quy** [comment]
Yes. To reproduce:

* Install staging.
* Log in on front end.
* Click `Article Category List`.
* Click `New` button.
* Click `Cancel` button.
* Redirects to home page.

On a site without `~`, it redirects to the article category list page.

**Comment by julienV** [comment]
you are right, the base64 return link on the 'new' button stll has a +

**Comment by julienV** [comment]
it's because of this:
https://github.com/joomla/joomla-cms/blob/staging/components/com_content/helpers/icon.php#L49

the return uri will be decoded when passing through JRoute::_() -> you get the +

**Comment by julienV** [comment]
I have tested this item :red_circle: unsuccessfully on 1e64a414c8566e5e43064669074e2130911c8832<br /><br />see PR comments, having issue with new button from icon content helper.
this might happen elsewhere as well...<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19196">issues.joomla.org/tracker/joomla-cms/19196</a>.</sub>

**Comment by csthomas** [comment]
Yes. 
There is a problem with `JRoute::()` which `decode` `%2B` back to `+`.
I can use `urlencode($return)` twice in `JRoute::()` or try to fix `JRoute::()` which is not easy and it will break unit tests.

This  create our problem:
https://github.com/joomla/joomla-cms/blob/7d3a3ef6c68be95baf4f822d339730e86454fbdd/libraries/vendor/joomla/uri/src/AbstractUri.php#L308-L311

**Comment by csthomas** [comment]
May be I should move the value of return to outside: `JRoute::_('...&return=') . urlencode($return)`

**Comment by julienV** [comment]
Indeed, that is what i wrote in the issue: the problem is that JRoute::_ calls UriHelper::parse_url() , which decodes the params.
I don't think double url_encode is a good solution, it's a hack.
Adding the return outside of JRoute::_ is better, but you need to make sure if there is already a ?var=xx in the returned url, as sef components could return an url without get variables.

**Comment by csthomas** [comment]
I fixed previous issues. Please test.

The best way would be to remove `urldecode` from:
https://github.com/joomla/joomla-cms/blob/7d3a3ef6c68be95baf4f822d339730e86454fbdd/libraries/vendor/joomla/uri/src/AbstractUri.php#L308-L311

and replace it by something better but this would break B/C.

File: administrator/components/com_menus/views/menus/tmpl/default.php [hunk 1]
```diff
@@ -171,7 +171,8 @@
 										<?php foreach ($this->modules[$item->menutype] as &$module) : ?>
 											<li>
 												<?php if ($user->authorise('core.edit', 'com_modules.module.' . (int) $module->id)) : ?>
-													<?php $link = JRoute::_('index.php?option=com_modules&task=module.edit&id=' . $module->id . '&return=' . $return . '&tmpl=component&layout=modal'); ?>
+													<?php $link = 'index.php?option=com_modules&task=module.edit&id=' . $module->id . '&tmpl=component&layout=modal&return=return'; ?>
```
Comment (Quy) [inline administrator/components/com_menus/views/menus/tmpl/default.php:174]: Can it be done this way to remove the use of `str_replace`?
```
<?php $link = 'index.php?option=com_modules&task=module.edit&id=' . $module->id . '&tmpl=component&layout=modal&return=' . urlencode($return); ?>
<?php $link = JRoute::_($link); ?>
```

Comment (csthomas) [reply administrator/components/com_menus/views/menus/tmpl/default.php:174]: This is the problem that in `JRoute::_()` this does not work.  
`JRoute::_()` at the end of own process uses `urldecode()` as I mentioned (`return urldecode(http_build_query($params, '', '&')); `) in my previous comment. 

The only way to encode the return value is to execute it outside the JRoute/JUri methods.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 43c20283319fdcdb128b1a2aa03e92caf39e0245<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19196">issues.joomla.org/tracker/joomla-cms/19196</a>.</sub>

**Comment by julienV** [comment]
This works, but it feels wrong... I agree with you  it's JRoute::_() that should be fixed. I am not sure why we urldecode is used there, maybe this needs to be forwarded to people that coded this initially, or have a better understanding of the router maybe (not saying you don't, but speaking for me)

**Comment by csthomas** [comment]
I tried to fix `JUri` and remove `urldecode` and decode only a few chars like ('[', ']') but it changes behaviour and unit tests fail. 

The issue was introduced at: 
https://github.com/joomla-framework/uri/blob/b91252fe255a7da9f8ab365750db693891d3324d/src/AbstractUri.php#L310

The last changes was done a long time ago.

I looked at drupal and wordpress:
- http://www.drupalcontrib.org/api/drupal/drupal%21core%21lib%21Drupal%21Component%21Utility%21Url.php/function/Url%3A%3AbuildQuery/8
- https://developer.wordpress.org/reference/functions/_http_build_query/

and they do not use `urldecode`. 

I'll put this on Glip's channel.

**Comment by csthomas** [comment]
7 days ago I asked for advice on Glip channel but there is no interest.

In this PR I use two solutions:

1. I used `str_replace('return=return', 'return=' . $return, $link)` because it is safe after `JRoute/Juri` will be patched.
2. I used `urlencode(urlencode($return))` where I do not have a full link but only partial query available.
   In that situation I added a comment because it is not safe after `JRoute/Juri` will be patched.
   In this situation I can not use point 1 because it breaks B/C.

Using method like `JRoute::_('...&return=') . urlencode($return)` is wrong too because `JRoute` may add a new parameter `&Itemid=...` and then we end with `&return=&Itemid=...`.

IMHO I could try to change `JUri::buildQuery` but it breaks B/C and has to wait till J4.

@julienV If you mark it as test success then maintainers will decide.

**Comment by csthomas** [comment]
Due to lack of interest, I close it.

**Pull request closed without merging**